### PR TITLE
Deprecate extracting() in favor of extractingByKey()/extractingByKeys() for MapAssert

### PR DESCRIPTION
**Pull request by scordio**
#### Check List:
* Fixes https://github.com/joel-costigliola/assertj-core/pull/1619#discussion_r328548320
* Unit tests : YES
* Javadoc with a code example (on API only) : YES

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java [hunk 1]
```diff
@@ -1566,6 +1569,40 @@ public SELF withThreadDumpOnError() {
...
+   * map.put("name", "kawhi");
+   * map.put("age", 25);
+   *
+   * assertThat(map).extractingByKeys("name", "age")
+   *                .contains("kawhi", 25);</code></pre>
+   * <p>
+   * Note that the order of extracted keys value is consistent with the iteration order of the array under test.
+   * <p>
+   * Nested keys are not yet supported, passing "name.first" won't get a value for "name" and then try to extract
+   * "first" from the previously extracted value, instead it will simply look for a value under "name.first" key.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: I'm not so sure if this section is valuable as it only applies for String keys, thoughts ?

File: src/test/java/org/assertj/core/api/BDDSoftAssertionsTest.java [hunk 2]
```diff
@@ -1522,8 +1522,7 @@ public void map_soft_assertions_should_report_errors_on_final_methods_and_method
     softly.then(map)
           .as("extracting(\"a\")")
           .overridingErrorMessage("error message")
-          // convert to Object otherwise will use extracting(String) in AbstractObjectAssert
-          .extracting((Object) "a")
+          .extractingByKey("a")
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/BDDSoftAssertionsTest.java]: that looks better indeed! :)

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java [hunk 3]
```diff
@@ -1566,6 +1569,40 @@ public SELF withThreadDumpOnError() {
...
+   * <p>
+   * Example:
+   * <pre><code class='java'> Map&lt;String, Object&gt; map = new HashMap&lt;&gt;();
+   * map.put("name", "kawhi");
+   * map.put("age", 25);
+   *
+   * assertThat(map).extractingByKeys("name", "age")
+   *                .contains("kawhi", 25);</code></pre>
+   * <p>
+   * Note that the order of extracted keys value is consistent with the iteration order of the array under test.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: Add another note to say that providing no keys will result in an empty list.

File: src/test/java/org/assertj/core/api/map/MapAssert_extractingByKeys_Test.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    then(error).hasMessage(actualIsNull());
+  }
+
+  @Test
+  void should_allow_list_assertions_on_values_extracted_from_given_map_keys() {
+    // WHEN
+    AbstractListAssert<?, List<?>, Object, ObjectAssert<Object>> result = assertThat(map).extractingByKeys(NAME, "age");
+    // THEN
+    result.containsExactly("kawhi", 25);
+  }
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/map/MapAssert_extractingByKeys_Test.java]: add a tests showing the behavior when providing no keys = will result in an empty list.

**Comment by joel-costigliola** [comment]
Good work, just a few comments!

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java (hunk 1 above)
Comment (scordio) [reply src/main/java/org/assertj/core/api/AbstractMapAssert.java]: I agree, it doesn't make much sense now. I was copy-pasting it from the corresponding `extracting` and even there it doesn't seem 100% right. The original source of this sentence was the javadoc for `extracting(Object... keys)` which has last change in 2017, probably the original concept has changed over time.

Also, this seems to overlap the description of `extracting` in `AbstractObjectAssert`:
https://github.com/joel-costigliola/assertj-core/blob/215c93d4b0ac3c040772ab552470e956d2cabb62/src/main/java/org/assertj/core/api/AbstractObjectAssert.java#L621

I'll remove this sentence from here.

**Comment by scordio** [comment]
@joel-costigliola PR ready!

**Comment by joel-costigliola** [comment]
Integrated thanks @scordio!

**Comment by joel-costigliola** [comment]
^^ Notice how I'm not putting a space before `!` 😆
